### PR TITLE
RE2022-218: parser refactor for microtrait

### PR DESCRIPTION
**Pull request by Tianhao-Gu**

### CODE REVIEW

File: src/loaders/genome_collection/parse_tool_results.py [hunk 1]
```diff
@@ -726,57 +442,50 @@ def microtrait(root_dir, env, kbase_collection, load_ver, fatal_ids):
...
+
+            # process heatmap metadata
+            with jsonlines.open(data_dir / MICROTRAIT_META, 'r') as jsonl_f:
+                metas = [meta for meta in jsonl_f]
+                if reference_meta is None:
+                    reference_meta = metas
+
+                # check if the metadata is consistent across all data dirs
+                if metas != reference_meta:
+                    raise ValueError(f'Inconsistent metadata for {data_dir}')
```
Comment (MrCreosote) [inline src/loaders/genome_collection/parse_tool_results.py:484]: There's only one line here for each genome, right? so
```suggestion
                meta = next(jsonl_f)
                if reference_meta is None:
                    reference_meta = meta

                # check if the metadata is consistent across all data dirs
                if meta != reference_meta:
                    raise ValueError(f'Inconsistent metadata for {data_dir}')
```

That confused the heck out of me

Comment (MrCreosote) [reply src/loaders/genome_collection/parse_tool_results.py:484]: I suppose you could also read the list, check that `len == 1`, and then `meta = metas[0]`

File: src/loaders/genome_collection/parse_tool_results.py [hunk 2]
```diff
@@ -475,247 +401,37 @@ def _create_meta_lookup(root_dir, env, kbase_collection, load_ver, tool):
...
+    # sort the columns by column id
+    for category in heatmap_categories.values():
+        category[FIELD_HEATMAP_COLUMNS] = sorted(category[FIELD_HEATMAP_COLUMNS],
+                                                 key=lambda x: int(x[FIELD_HEATMAP_COL_ID]))
 
-    if _is_binary(num):
-        return bool(num)
-    else:
-        raise ValueError(f'Input must be a binary number (i.e., 0 or 1). Got {num} instead.')
+    heatmap_meta = {FIELD_HEATMAP_CATEGORIES: list(heatmap_categories.values()),
```
Comment (MrCreosote) [inline src/loaders/genome_collection/parse_tool_results.py:427]: Aren't the categories supposed to be sorted by the column ID?

File: src/loaders/genome_collection/parse_tool_results.py [hunk 3]
```diff
@@ -475,247 +401,37 @@ def _create_meta_lookup(root_dir, env, kbase_collection, load_ver, tool):
...
+        meta_without_category = meta.copy()
+        del meta_without_category[FIELD_HEATMAP_CATEGORY]
+        heatmap_categories[category][FIELD_HEATMAP_COLUMNS].append(meta_without_category)
 
-def _num_to_bool(num: int | bool) -> bool:
-    # Given a number, checks if it is a binary type (i.e., contains only 0s and 1s).
+    # sort the columns by column id
+    for category in heatmap_categories.values():
+        category[FIELD_HEATMAP_COLUMNS] = sorted(category[FIELD_HEATMAP_COLUMNS],
+                                                 key=lambda x: int(x[FIELD_HEATMAP_COL_ID]))
```
Comment (MrCreosote) [inline src/loaders/genome_collection/parse_tool_results.py:425]: We do an equivalent sort for the rows in the tool container right?

File: src/loaders/genome_collection/parse_tool_results.py [hunk 4]
```diff
@@ -475,247 +401,37 @@ def _create_meta_lookup(root_dir, env, kbase_collection, load_ver, tool):
...
-    # sort ColumnCategory objects by the column id of the first column in each ColumnCategory object
-    sorted_categories = sorted(
-        categories.values(), key=lambda category: int(category.columns[0].col_id)
-    )
-
-    # this is to ensure that the column ids are in ascending order
-    # TODO: might need to skip this step for heatmap products other than microtrait
-    column_ids = [column.col_id for category in sorted_categories for column in category.columns]
-    if not _ensure_list_ordered(column_ids):
-        raise ValueError(f'Column ids are not ordered in ascending order: {column_ids}')
```
Comment (MrCreosote) [inline src/loaders/genome_collection/parse_tool_results.py:548]: Do we no longer need this check?

File: src/loaders/genome_collection/parse_tool_results.py [hunk 5]
```diff
@@ -475,247 +401,37 @@ def _create_meta_lookup(root_dir, env, kbase_collection, load_ver, tool):
...
-        detected_genes_score: dict[str, float] = None,
-) -> (float | int, float | int, CellDetail):
-    # Append a cell to the heatmap row and return the global min and max values
-
-    if trait_type == 'count':
-        trait_val = _int(trait_val)
-    elif trait_type == 'binary':
-        trait_val = _num_to_bool(trait_val)
-    else:
-        raise ValueError(f'Unknown trait type {trait_type}')
```
Comment (MrCreosote) [inline src/loaders/genome_collection/parse_tool_results.py:613]: Does this conversion need to be done in the tool container? I don't remember seeing anything like this

File: src/common/product_models/heatmap_common_models.py [hunk 6]
```diff
@@ -22,6 +22,8 @@
 FIELD_HEATMAP_DESCR = "description"
 FIELD_HEATMAP_TYPE = "type"
 FIELD_HEATMAP_CATEGORY = "category"
+FIELD_HEATMAP_COLUMNS = "columns"
+FIELD_HEATMAP_CATEGORIES = "categories"
 
 _FLD_CELL_ID = Field(
     example="4",
```
[the remark this answers is not in the document]
Comment (Tianhao-Gu) [reply src/common/product_models/heatmap_common_models.py:29]: hmmm. I don't think so.

File: src/loaders/genome_collection/parse_tool_results.py [hunk 7]
```diff
@@ -44,26 +44,24 @@
 import shutil
```
[the remark this answers is not in the document]
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_tool_results.py:44]: I think so. 

same samples:
`kbcoll_microtrait_cells.jsonl`:
```
{"cell_id": "2a8b0877-c7b2-4126-8c01-7169e983eeec", "values": {}, "coll": "GROW", "load_ver": "tian_test_podman", "_key": "2a8b0877-c7b2-4126-8c01-7169e983eeec"}
{"cell_id": "7683d6f6-9481-4ce5-883c-6ca3ce8f36de", "values": {"exbD": 94.0, "exbB": 197.1}, "coll": "GROW", "load_ver": "tian_test_podman", "_key": "7683d6f6-9481-4ce5-883c-6ca3ce8f36de"}
```

`kbcoll_microtrait_data.jsonl`:
```
{"kbase_id": "69037_4_1", "cells": [{"cell_id": "2a8b0877-c7b2-4126-8c01-7169e983eeec", "col_id": "1", "val": 0}, {"cell_id": "7683d6f6-9481-4ce5-883c-6ca3ce8f36de", "col_id": "2", "val": 2}, {"cell_id": "bff4fc4a-2bdb-4e7d-995a-2fa600156d07", "col_id": "3", "val": 0}, {"cell_id": "4898e03d-80a6-412f-8bf0-a65fa8c4e502", "col_id": "4", "val": 0}, {"cell_id": "2a6211a6-7c27-4f7a-964f-392599e00a69",
...
```

`kbcoll_microtrait_meta.jsonl`:
```
{"categories": [{"category": "Resource Acquisition", "columns": [{"col_id": "1", "name": "Aromatic acid transport", "description": "Resource Acquisition:Substrate uptake:aromatic acid transport", "type": "count"}, {"col_id": "2", "name": "Biopolymer transport", "description": "Resource Acquisition:Substrate uptake:biopolymer transport", "type": "count"}, {"col_id": "3", "name": "Amino sugar transport", "description": "Resource Acquisition:Substrate uptake:carbohydrate transport:amino sugar transport", "type": "count"}, {"col_id": "4", "name": "Carbohydrate acid transport", 
...
```

File: src/loaders/genome_collection/parse_tool_results.py [hunk 8]
```diff
@@ -44,26 +44,24 @@
...
 import sys
```
[the remark this answers is not in the document]
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_tool_results.py:45]: after addressing comments, I don't think we are missing anything.

File: src/loaders/genome_collection/parse_tool_results.py [hunk 9]
```diff
@@ -475,247 +401,37 @@ def _create_meta_lookup(root_dir, env, kbase_collection, load_ver, tool):
...
-    # loop over each trait in traits_meta
-    for trait in traits_meta.values():
-        trait_category = trait[_SYS_TRAIT_CATEGORY]
-
-        if trait[_SYS_TRAIT_TYPE] == 'count':
-            trait_type = ColumnType.COUNT.value
-        elif trait[_SYS_TRAIT_TYPE] == 'binary':
-            trait_type = ColumnType.BOOL.value
-        else:
-            raise ValueError(f'Unknown trait type {trait[_SYS_TRAIT_TYPE]}')
```
[the remark this answers is not in the document]
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_tool_results.py:521]: 👍 Added in microtait tool container

File: src/loaders/genome_collection/parse_tool_results.py (hunk 2 above)
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_tool_results.py]: 👍 added

File: src/loaders/genome_collection/parse_tool_results.py (hunk 3 above)
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_tool_results.py]: no. But It's good idea to sort it there. I added the sort in tool runner.

File: src/loaders/genome_collection/parse_tool_results.py (hunk 4 above)
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_tool_results.py:548]: I think we explicitly sort. Maybe we don't need to check again.

File: src/loaders/genome_collection/parse_tool_results.py (hunk 5 above)
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_tool_results.py:613]: 👍 added.

File: src/loaders/genome_collection/parse_tool_results.py (hunk 1 above)
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_tool_results.py]: for each genome, we create `microtrait_meta.jsonl` something like:
```
{"col_id": "1", "name": "Aromatic acid transport", "description": "Resource Acquisition:Substrate uptake:aromatic acid transport", "category": "Resource Acquisition", "type": "count"}
{"col_id": "2", "name": "Biopolymer transport", "description": "Resource Acquisition:Substrate uptake:biopolymer transport", "category": "Resource Acquisition", "type": "count"}
{"col_id": "3", "name": "Amino sugar transport", "description": "Resource Acquisition:Substrate uptake:carbohydrate transport:amino sugar transport", "category": "Resource Acquisition", "type": "count"}
...
```
I want to read the meta_data from the first directory. And see if there are any different metadata from the rest dirs.

File: src/common/product_models/heatmap_common_models.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply src/common/product_models/heatmap_common_models.py:29]: I was thinking of this https://github.com/kbase/collections/pull/419#discussion_r1296664374

File: src/loaders/genome_collection/parse_tool_results.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_tool_results.py:44]: Sorry, are those the same as the old output? That's what I think you're saying but I'm not sure

File: src/loaders/genome_collection/parse_tool_results.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_tool_results.py:45]: :+1:

File: src/loaders/genome_collection/parse_tool_results.py (hunk 1 above)
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_tool_results.py]: Oh I see, for some reason I thought that it'd look like the metadata model. So 1 line per column, got it.

Could you add a brief comment explaining that?

File: src/loaders/genome_collection/parse_tool_results.py (hunk 4 above)
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_tool_results.py:548]: I think the check is affirming that the column IDs are ordered through the categories as well e.g it doesn't look like

category 1
   4
   5
   6
category 2
   1
   2
   3

If you tell me that can't happen that's fine with me though

File: src/common/product_models/heatmap_common_models.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Tianhao-Gu) [reply src/common/product_models/heatmap_common_models.py:29]: Oh, they are meant to be removed. They are removed from this PR.

File: src/loaders/genome_collection/parse_tool_results.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_tool_results.py:44]: I think they are exactly the same.

File: src/loaders/genome_collection/parse_tool_results.py (hunk 1 above)
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_tool_results.py]: 👍 added.

File: src/loaders/compute_tools/microtrait/env_microtrait.yaml [hunk 10]
```diff
@@ -39,4 +39,4 @@ dependencies:
   - pandas=2.0.0
   - unzip=6.0
   - jsonlines=3.1.0
-  - pydantic=1.10.12
```
Comment (Tianhao-Gu) [inline src/loaders/compute_tools/microtrait/env_microtrait.yaml:42]: `import heatmap_common_models` with pydantic 1 does break.

Comment (MrCreosote) [reply src/loaders/compute_tools/microtrait/env_microtrait.yaml:42]: That's surprising. What's the error?

Comment (Tianhao-Gu) [reply src/loaders/compute_tools/microtrait/env_microtrait.yaml:42]: ah oh. cleared the terminal. But I recall trackback is from `heatmap_common_models`.

Comment (MrCreosote) [reply src/loaders/compute_tools/microtrait/env_microtrait.yaml:42]: Weird, I can't understand how those changes could make pydantic 1 puke

File: src/loaders/genome_collection/parse_tool_results.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_tool_results.py:44]: :+1:

File: src/loaders/genome_collection/parse_tool_results.py (hunk 5 above)
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_tool_results.py:613]: :+1:

File: src/loaders/genome_collection/parse_tool_results.py (hunk 3 above)
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_tool_results.py]: :+1:

File: src/loaders/genome_collection/parse_tool_results.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_tool_results.py:521]: :+1:

File: src/loaders/genome_collection/parse_tool_results.py (hunk 2 above)
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_tool_results.py]: :+1:

File: src/loaders/genome_collection/parse_tool_results.py (hunk 3 above)
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_tool_results.py]: It looks like you took out the sort in the parser, are you just relying on the tool runner sort to make sure the meta columns are ordered?

File: src/loaders/genome_collection/parse_tool_results.py (hunk 4 above)
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_tool_results.py:548]: sure. Added the check.

File: src/loaders/genome_collection/parse_tool_results.py (hunk 3 above)
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_tool_results.py]: yea. I think it makes sense to sort the pre-processed results in tool runner. Do you think it's necessary to double check in the global parser?

File: src/loaders/genome_collection/parse_tool_results.py (hunk 1 above)
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_tool_results.py]: I was thinking of something like 

```
# The microtrait runner saves the metadata as jsonl with 1 trait column per line
metas = [meta for meta in jsonl_f]
```

Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_tool_results.py]: 👍

File: src/common/product_models/heatmap_common_models.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply src/common/product_models/heatmap_common_models.py:29]: :+1:

File: src/loaders/genome_collection/parse_tool_results.py (hunk 1 above)
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_tool_results.py]: :+1:

File: src/loaders/genome_collection/parse_tool_results.py (hunk 4 above)
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_tool_results.py:548]: :+1:

File: src/loaders/genome_collection/parse_tool_results.py (hunk 3 above)
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_tool_results.py]: Nope, that seems fine, I just wanted to make sure I understood the change

:+1:

**Review by MrCreosote — APPROVED** [review]
:LGTM

**Pull request merged**